pkg/chains/signing.go

```diff
@@ -252,10 +282,34 @@ func (ts *TaskRunSigner) SignTaskRun(ctx context.Context, tr *v1beta1.TaskRun) e
 			}
 		}
 		if merr.ErrorOrNil() != nil {
+			if err := HandleRetries(tr, ts, extraAnnotations); err != nil {
+				merr = multierror.Append(merr, err)
+			}
 			return merr
 		}
 	}
 
 	// Now mark the TaskRun as signed
-	return MarkSigned(tr, ts.Pipelineclientset, extraAnnotations)
+	return MarkSigned(tr, ts.Pipelineclientset, extraAnnotations, "true")
+}
+
+// HandleRetries adds retry annotation to keep track of retries and marks the job as failed
+// if retries have reached MaxRetries.
+func HandleRetries(tr *v1beta1.TaskRun, ts *TaskRunSigner, extraAnnotations map[string]string) error {
+	ok, s := RetryAgain(tr)
+	if !ok {
+		return AddRetries(tr, ts.Pipelineclientset, extraAnnotations, 1)
+	}
+
+	count, err := strconv.Atoi(s)
+	if err != nil {
+		return err
+	}
+
+	if count <= MaxRetries {
+		return AddRetries(tr, ts.Pipelineclientset, extraAnnotations, count+1)
+	}
+
+	// Marking it failed after retrying MaxRetries
+	return MarkSigned(tr, ts.Pipelineclientset, extraAnnotations, Failed)
```

Comment:
I think it would be more readable if we added a `MarkFailed` function and used that instead

Reply:
Addressed it. Thanks